django/db/models/expressions.py

```diff
@@ -1367,11 +1367,20 @@ def resolve_expression(
         c = self.copy()
         c.is_summary = summarize
         if hasattr(c.condition, "resolve_expression"):
+            extra_args = (
+                dict(reuse_with_filtered_relation=True)
+                if isinstance(c.condition, Q)
+                else {}
+            )
             c.condition = c.condition.resolve_expression(
-                query, allow_joins, reuse, summarize, False
+                query, allow_joins, reuse, summarize, False, **extra_args
```

Comment:
I wonder why we only pass `reuse_with_filtered_relation` here :thinking:. It seems that we're only fixing an issue in `When()`, but in theory I could appear in other places. This looks especially suspicious since we're not even aware that this `When()` is used with `FilteredRelation()` or not. Maybe we always should pass `reuse_with_filtered_relation=True` to `_add_q()` when resolving `Q` objects :thinking: (\cc @charettes).

Reply:
I agree, I've been mislead because `When()` is the only expression in tests having a `Q()` object as child.

I like this approach, cleaner and more generic